aten/src/ATen/native/cpu/UpSampleKernel.cpp

```diff
@@ -573,16 +573,26 @@ struct HelperInterpNearest : public HelperInterpBase {
     AT_DISPATCH_FLOATING_TYPES(
       scalar_type, "compute_indices_weights_nearest", [&] {
 
-        scalar_t scale = area_pixel_compute_scale<scalar_t>(input_size, output_size, align_corners, opt_scale);
-
         auto input_index_ptr = output[0].data_ptr<int64_t>();
-        int64_t input_index;
-
-        for (int64_t i=0; i<output_size; i++) {
-          const scalar_t real_input_index = area_pixel_compute_source_index<scalar_t>(
-              scale, i, /*align_corners=*/true, /*cubic=*/false);
-          input_index = static_cast<int64_t>(floorf(real_input_index));
-          input_index_ptr[i] = static_cast<int64_t>(std::min(input_index, input_size - 1)) * stride;
+        if (output_size == input_size) {
```

Comment:
Do you know whether removing these special cases in the previous implementation would still work? These seem like "fast paths" and it is worrisome to me that the generic else case wouldn't work correctly without these special cases. Can you please verify this?

Reply:
If you mean to comment out this code : https://github.com/pytorch/pytorch/blob/56b43f4fec1f76953f15a627694d4bba34588969/aten/src/ATen/native/cpu/UpSampleKernel.cpp#L18-L24 in v1.8.0, then once recompiled it produces the same output as 1.9.0:
```python
import torch
import torch.nn.functional as F
print("pytorch version:", torch.__version__)

t_in = torch.tensor([[[1.0, 2.0, 1.0]]])
scale_factor = 1.00001
t_out = F.interpolate(t_in, scale_factor=scale_factor, recompute_scale_factor=False, mode="nearest")

print("input:", t_in)
print("output:", t_out)
```
>
```
pytorch version: 1.9.0
input: tensor([[[1., 2., 1.]]])
output: tensor([[[1., 1., 2.]]])
```
and
```
pytorch version: 1.8.0a0+0e80d4f
input: tensor([[[1., 2., 1.]]])
output: tensor([[[1., 1., 2.]]])
```

The main question however is what would be correct/expected result ? According to https://ppwwyyxx.com/blog/2021/Where-are-Pixels/ (cc @ppwwyyxx ), scikit-image has bug-free `resize` implementation. For our case, similar scikit-image op is [`rescale`](https://github.com/scikit-image/scikit-image/blob/e128e0743bb2cded104654352a13ed5b4e53efb8/skimage/transform/_warps.py#L201) which produces:
```
skimage: 0.18.2
input: [[1. 2. 1.]]
scale: 1.00001
output: [[1. 2. 1.]]
```
and this actually corresponds to `recompute_scale_factor = True`. PyTorch v1.9.0 and v1.8.0 both gives the same result if using `recompute_scale_factor = True`.

Reply:
The special case works differently from the generic case likely because the generic case is incorrect (and there is a bug tracking this: https://github.com/pytorch/pytorch/issues/34808)

Reply:
That's true that generic case has also an issue, but I think it is not related here. Currently, ` F.interpolate(t_in, scale_factor=scale_factor, recompute_scale_factor=False, mode="nearest")` is similar to:
```python
import torch

s = 10
t_in = torch.arange(s, dtype=torch.float).unsqueeze(0).unsqueeze(0)
scale = 1.00001
half_pixel = False

output_size = (1, 1, s)
resized = torch.zeros(output_size)
if scale is None:
    # equivalent to recompute_scale_factor = True and skimage
    scales = (1, 1, t_in.shape[2] / resized.shape[2])
else:
    scales = (1, 1, 1.0 / scale)

for i in range(resized.shape[2]):
    if half_pixel:
        x = (i + 0.5) * scales[2] - 0.5
    else:
        x = i * scales[2]
    ii = int(x)
    resized[0, 0, i] = t_in[0, 0, ii]
    
print(resized)
> scale = 1.0001, half_pixel = False
> tensor([[[0., 0., 1., 2., 3., 4., 5., 6., 7., 8.]]])
> scale = 1.0001, half_pixel = True
> tensor([[[0., 0., 1., 2., 3., 4., 5., 6., 7., 8.]]])
> scale = None, half_pixel = False
> tensor([[[0., 1., 2., 3., 4., 5., 6., 7., 8., 9.]]])
> scale = None, half_pixel = True
> tensor([[[0., 1., 2., 3., 4., 5., 6., 7., 8., 9.]]])
```